storage/operation/reads.go

```diff
@@ -0,0 +1,207 @@
+					return fmt.Errorf("could not handle entity: %w", err)
+				}
+
+				return nil
+			})
+
+			if err != nil {
+				return fmt.Errorf("could not process value: %w", err)
+			}
+		}
+
+		return nil
+	}
+}
+
+// Traverse will iterate over all keys with the given prefix
+func Traverse(prefix []byte, iterFunc IterationFunc, opt storage.IteratorOption) func(storage.Reader) error {
+	return Iterate(prefix, prefix, iterFunc, opt)
+}
+
+// Exists returns true if a key exists in the database.
+// No errors are expected during normal operation.
+func Exists(key []byte, keyExists *bool) func(storage.Reader) error {
+	return func(r storage.Reader) error {
+		_, closer, err := r.Get(key)
+		if err != nil {
+			// the key does not exist in the database
+			if errors.Is(err, storage.ErrNotFound) {
+				*keyExists = false
+				return nil
+			}
+			// exception while checking for the key
+			return irrecoverable.NewExceptionf("could not load data: %w", err)
+		}
+		defer closer.Close()
+
+		// the key does exist in the database
+		*keyExists = true
+		return nil
+	}
+}
+
+// Retrieve will retrieve the binary data under the given key from the database
+// and decode it into the given entity. The provided entity needs to be a
+// pointer to an initialized entity of the correct type.
+// Error returns:
+//   - storage.ErrNotFound if the key does not exist in the database
+//   - generic error in case of unexpected failure from the database layer, or failure
+//     to decode an existing database value
+func Retrieve(key []byte, entity interface{}) func(storage.Reader) error {
+	return func(r storage.Reader) error {
+		val, closer, err := r.Get(key)
+		if err != nil {
+			return err
+		}
+
+		defer closer.Close()
+
+		err = msgpack.Unmarshal(val, entity)
+		if err != nil {
+			return irrecoverable.NewExceptionf("could not decode entity: %w", err)
+		}
+		return nil
+	}
+}
+
+// FindHighestAtOrBelow finds the highest key with the given prefix and
+// height equal to or below the given height.
+func FindHighestAtOrBelow(prefix []byte, height uint64, entity interface{}) func(storage.Reader) error {
+	return func(r storage.Reader) error {
+		if len(prefix) == 0 {
+			return fmt.Errorf("prefix must not be empty")
+		}
+
+		key := append(prefix, EncodeKeyPart(height)...)
+		it, err := r.NewIter(prefix, key, storage.DefaultIteratorOptions())
+		if err != nil {
+			return fmt.Errorf("can not create iterator: %w", err)
+		}
+		defer it.Close()
+
+		var highestKey []byte
+		// find highest value below the given height
+		for it.First(); it.Valid(); it.Next() {
+			highestKey = it.IterItem().Key()
+		}
+
+		if len(highestKey) == 0 {
+			return storage.ErrNotFound
+		}
+
+		// read the value of the highest key
+		val, closer, err := r.Get(highestKey)
+		if err != nil {
+			return err
+		}
+
+		defer closer.Close()
+
+		err = msgpack.Unmarshal(val, entity)
+		if err != nil {
+			return irrecoverable.NewExceptionf("failed to decode value: %w", err)
+		}
+
+		return nil
+	}
+}
```

Comment:
Yeah, exactly like Jordan said (thanks for explaining it), that it minimizes the changes when refactoring existing modules, so that the changes are easier for us to review. 

However, if we take a step back and rethink about why we need these wraps original, I actually agree with Alex's point that the functors (Get, Retrieve, etc) are unnecessary wraps if we are not using transactions. 

So I refactored them to keep both the implementation by extracting the non-functor version, and moved the functor versions into the `read_functors.go` files, marked them to be deprecated as we are gonna replace stateful transactions with stateless batch updates. Reasonings are added as comments in that file.

Thanks for the idea Alex.